inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/InstrumentationManager.java

```diff
@@ -0,0 +1,241 @@
+                }
+            }
+        });
+        instrumentationThread.setName("inspectit-instrumenter");
+        instrumentationThread.setDaemon(true);
+        instrumentationThread.start();
+    }
+
+    private void applyInstrumentation(InstrumentationSettings instrConf, boolean batching) {
+
+        appliedInstrumentation = instrConf;
+        try (val s = selfMonitoring.withSelfMonitoring(getClass().getSimpleName() + "-instrumentation")) {
+
+            log.info("Performing instrumentation...");
+            AgentBuilder baseAgent = new AgentBuilder.Default().disableClassFormatChanges();
+            if (batching) {
+                baseAgent = baseAgent.with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
+                        .with(AgentBuilder.RedefinitionStrategy.BatchAllocator.ForFixedSize.ofSize(instrConf.getBatchSize()))
+                        .with(AgentBuilder.RedefinitionStrategy.Listener.Pausing.of(instrConf.getInterBatchPause().toMillis(), TimeUnit.MILLISECONDS));
+            } else {
+                baseAgent = baseAgent.with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION);
+            }
+            baseAgent = setupIgnoredClasses(baseAgent)
+                    .with(new AgentBuilder.Listener.Adapter() {
+
+                        @Override
+                        public void onError(String typeName, ClassLoader classLoader, JavaModule module, boolean loaded, Throwable throwable) {
+                            log.error("Error transforming {} from class loader {}", typeName, classLoader, throwable);
+                        }
+
+                        @Override
+                        public void onTransformation(TypeDescription typeDescription, ClassLoader classLoader, JavaModule module, boolean loaded, DynamicType dynamicType) {
+                            log.debug("Starting transformation of {}", typeDescription.getName());
+                        }
+                    });
+
+            AgentBuilder agent = baseAgent;
+            for (val sensor : specialSensors) {
+                if (sensor.isEnabledForConfig(instrConf)) {
+                    agent = sensor.instrument(instrConf, agent);
+                }
+            }
+            previousInstrumentation = Optional.of(agent.installOn(instrumentation));
+        }
+        log.info("Instrumentation is done.");
+    }
+
+    /**
+     * Unfortunately issuing multiple AgentBuilder#ignore calls only takes the last one.
+     * Therefore we have to build a custom matcher.
+     */
+    private AgentBuilder setupIgnoredClasses(AgentBuilder baseAgent) {
+
+        ClassLoader inspectITClassLoader = getClass().getClassLoader();
+
+        return baseAgent.ignore((type, classloader, module, clazz, pd) -> {
+            if (classloader == inspectITClassLoader) {
+                return true;
+            } else if (classloader == null) {
+                return IGNORED_BOOTSTRAP_PACKAGES.stream()
```

Comment:
This is not enough, we need to check if the class was loaded by the bootstrap class loader, otherwise, you are excluding all classes from given packages, even if they are loaded by the application class loader.. That seems not correct imo.. Or